### PR TITLE
Raw forms, Errors Duplicates ES

### PR DESCRIPTION
**Pull request by benrudolph**
@snopoke This converts the Raw Forms report to ES. 

screen shot 2016-02-18 at 5 57 00 pm

One caveat is that we don't send all the doc types to ES which makes the report not as comprehensive. I've updated the couch filter so that we do, but I'm not totally confident there will be no side effects

### CODE REVIEW

**Comment by benrudolph** [comment]
cc: @gcapalbo 

https://github.com/dimagi/dimagi-utils/pull/349

File: corehq/apps/reports/standard/forms/reports.py [hunk 1]
```diff
@@ -95,12 +78,11 @@ def rows(self):
         EMPTY_USER = _("No User")
         EMPTY_FORM = _("Unknown Form")
 
-        items = self.paginator_results.get(self.pagination.start, self.pagination.count)
-        
-        def _to_row(error_doc):
+
+        def _to_row(xform_dict):
```
Comment (dimagimon) [inline corehq/apps/reports/standard/forms/reports.py]: E303 too many blank lines (2)

**Comment by snopoke** [comment]
:boot: :construction:

File: corehq/apps/reports/standard/forms/reports.py [hunk 2]
```diff
@@ -109,16 +90,26 @@ def _fmt_url(doc_id):
...
+                form_name = xmlns_to_name(
+                    self.domain,
+                    xform_dict.get('xmlns'),
+                    app_id=xform_dict.get('app_id'),
+                )
+            else:
+                form_name = EMPTY_FORM
+            return [
+                _fmt_url(xform_dict['_id']),
+                xform_dict['form']['meta']['username'] if xform_dict['form'].get('meta') else EMPTY_USER,
```
Comment (snopoke) [inline corehq/apps/reports/standard/forms/reports.py]: seems like very similar logic to `form_name`, could use the same if else clause and make this list construction even tidier

File: corehq/ex-submodules/couchforms/_design/filters/xforms.js [hunk 3]
```diff
@@ -11,10 +11,12 @@ function(doc, req)
...
         case "XFormArchived":
+        case "XFormError":
         case "XFormDeprecated":
         case "XFormDuplicate":
         case "XFormInstance-Deleted":
         case "HQSubmission":
+        case "SubmissionErrorLog":
             return true;
         default:
             return false;
```
Comment (snopoke) [inline corehq/ex-submodules/couchforms/_design/filters/xforms.js]: This is going to be a big reindex. Another option could be to create a new couchapp with just this filter

**Comment by snopoke** [comment]
Looks good. Added reindex label. Going to wait till we can chat about the reindex vs new couchapp.

File: corehq/ex-submodules/couchforms/_design/filters/xforms.js (hunk 3 above)
Comment (czue) [inline corehq/ex-submodules/couchforms/_design/filters/xforms.js]: I could be wrong, but I think that filter changes actually don't require any indexing. pretty sure @dannyroberts or @esoergel looked into that recently

**Comment by czue** [comment]
will be sad (but I guess necessary?) to get this report on ES. it's the one surefire place you can still find out if something is in the primary DB. @benrudolph did you consider implementing through the form processor interface? I think we'd have all the information necessary to do that.

File: corehq/ex-submodules/couchforms/_design/filters/xforms.js (hunk 3 above)
Comment (esoergel) [inline corehq/ex-submodules/couchforms/_design/filters/xforms.js]: That sounds right, but I don't remember for sure.  @dannyroberts would know.

**Comment by snopoke** [comment]
@czue good point

File: corehq/apps/es/forms.py [hunk 4]
```diff
@@ -54,3 +55,7 @@ def completed(gt=None, gte=None, lt=None, lte=None):
 
 def user_id(user_ids):
     return filters.term('form.meta.userID', list(user_ids))
+
+
+def doc_types(doc_types):
+    return filters.term('doc_type', map(lambda doc_type: doc_type.lower(), list(doc_types)))
```
Comment (esoergel) [inline corehq/apps/es/forms.py]: There's already a [`doc_type` filter](https://github.com/dimagi/commcare-hq/blob/cb3c4b0aa482d3d1bcf99b804ac8a5e4c4d26f86/corehq/apps/es/filters.py#L77-L79) on `HQESQuery` which accepts a list.  It doesn't coerce to lowercase though, so might have to push that to the client code or something if it's necessary.

**Comment by benrudolph** [comment]
@czue will try and take a stab at form processor. though it will be significantly uglier, ES made this whole report a ton cleaner

File: corehq/apps/es/forms.py (hunk 4 above)
Comment (benrudolph) [inline corehq/apps/es/forms.py]: :+1: thanks that makes sense

**Comment by czue** [comment]
cool. like i said not sure it's worth it so feel free to punt if it feels painful

**Comment by benrudolph** [comment]
hmm ok, i think let's try out this ES version since it simplifies the code a lot. if it turns out that we actually do need the SQL/Couch source, happy to implement

File: corehq/apps/es/forms.py (hunk 4 above)
Comment (benrudolph) [inline corehq/apps/es/forms.py]: https://github.com/dimagi/commcare-hq/commit/0607641f51d6d0e72fa4953a0bcb60f7a355e65a

File: corehq/apps/reports/standard/forms/reports.py (hunk 2 above)
Comment (benrudolph) [inline corehq/apps/reports/standard/forms/reports.py]: so much nicer: https://github.com/dimagi/commcare-hq/commit/0607641f51d6d0e72fa4953a0bcb60f7a355e65a

File: corehq/ex-submodules/couchforms/_design/filters/xforms.js (hunk 3 above)
Comment (dannyroberts) [inline corehq/ex-submodules/couchforms/_design/filters/xforms.js]: I would guess that what @czue said is right, but I actually haven't ever tested that. It'd be pretty easy to throw this up on staging and see if it causes a reindex though! (Even the amount of data we have on staging is usually enough to cause a > 5-min preindex.)

Comment (benrudolph) [inline corehq/ex-submodules/couchforms/_design/filters/xforms.js]: thanks will give that a go

**Comment by benrudolph** [comment]
(this did not cause a reindex)

**Comment by czue** [comment]
:relieved:

**Comment by benrudolph** [comment]
test failure unrelated, but realized i haven't tested this with the extra

**Comment by benrudolph** [comment]
:white_check_mark:

**Comment by czue** [comment]
:meat_on_bone: :page_facing_up: :rocket: :mag:

**Pull request merged**